File: src/eQuantic.UI.Web/ThemeBridge.cs
using System.Globalization;
using System.Text;
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Web;

/// <summary>
/// Serializes any <see cref="IAppTheme"/> to the compact JSON the browser runtime rehydrates into an
/// <c>AppTheme</c> (see <c>materializeTheme</c> in <c>theme-bridge.ts</c>) — the SSR→client THEME
/// BRIDGE. The server emits this blob into <c>window.__EQ_THEME__</c> next to <c>__EQ_CONFIG</c>, and
/// boot calls <c>setPhotonTheme</c> with it before hydration, so client re-renders resolve the SAME
/// colors/shape the SSR markup baked in (both sides read resolved values, not CSS vars — see
/// <see cref="TokenCss.Value"/>). The wire shape mirrors the <c>DesignSystemTsGenerator</c>'s
/// <c>photonTheme</c> exactly; the round-trip is cross-pinned in vitest against <c>photonTheme</c>.
/// </summary>
public static class ThemeBridge
{
    /// <summary>The theme as a single-line JSON object: surfaces, disabledOpacity, per-variant color
    /// groups, the type scale, elevations 0–5 and the shape scale — colors as <c>[[r,g,b,a],[r,g,b,a]]</c>
    /// light/dark tuples, keys camelCased to match the enum-lowered names the client looks up.</summary>
    public static string SerializeJson(IAppTheme theme)
    {
        var sb = new StringBuilder();
        sb.Append('{');

        sb.Append("\"surfaces\":{");
        var first = true;
        foreach (var property in typeof(IAppTheme).GetProperties()
                     .Where(p => p.PropertyType == typeof(ColorToken)))
        {
            Comma(sb, ref first);
            Key(sb, Camel(property.Name));
            Token(sb, (ColorToken)property.GetValue(theme)!);
        }
        sb.Append('}');

        sb.Append(",\"disabledOpacity\":").Append(Num(theme.DisabledOpacity));

        sb.Append(",\"variants\":{");
        first = true;
        foreach (var variant in Enum.GetValues<Variant>())
        {
            Comma(sb, ref first);
            Key(sb, Camel(variant.ToString()));
            var c = theme.Colors(variant);
            sb.Append('[');
            Token(sb, c.Base); sb.Append(',');
            Token(sb, c.OnBase); sb.Append(',');
            Token(sb, c.Pressed); sb.Append(',');
            Token(sb, c.Subtle); sb.Append(',');
            Token(sb, c.OnSubtle);
            sb.Append(']');
        }
        sb.Append('}');

        sb.Append(",\"type\":{");
        first = true;
        foreach (var role in Enum.GetValues<TypeRole>())
        {
            Comma(sb, ref first);
            Key(sb, Camel(role.ToString()));
            var s = theme.Type(role);
            sb.Append('[').Append(Num(s.Size)).Append(',').Append(Num(s.LineHeight))
              .Append(",\"").Append(Camel(s.Weight.ToString())).Append("\",")
              .Append(Num(s.Tracking)).Append(',').Append(Num(s.MaxScale)).Append(']');
        }
        sb.Append('}');

        sb.Append(",\"elevations\":[");
        for (var level = 0; level <= 5; level++)
        {
            if (level > 0) sb.Append(',');
            var e = theme.Elevation(level);
            sb.Append('[').Append(Num(e.OffsetY)).Append(',').Append(Num(e.Blur)).Append(',')
              .Append(Num(e.Spread)).Append(',');
            Token(sb, e.Color);
            sb.Append(']');
        }
        sb.Append(']');

        sb.Append(",\"shape\":{");
        first = true;
        foreach (var scale in Enum.GetValues<ShapeScale>())
        {
            Comma(sb, ref first);
            Key(sb, Camel(scale.ToString()));
            sb.Append(Num(theme.Shape(scale)));
        }
        sb.Append('}');

        sb.Append('}');
        return sb.ToString();
    }

    private static void Token(StringBuilder sb, ColorToken token)
    {
        sb.Append('[');
        Color(sb, token.Light);
        sb.Append(',');
        Color(sb, token.Dark);
        sb.Append(']');
    }

    private static void Color(StringBuilder sb, Color c) => sb
        .Append('[').Append(c.R).Append(',').Append(c.G).Append(',').Append(c.B).Append(',').Append(c.A).Append(']');

    private static void Key(StringBuilder sb, string name) => sb.Append('"').Append(name).Append("\":");

    private static void Comma(StringBuilder sb, ref bool first)
    {
        if (!first) sb.Append(',');
        first = false;
    }

    private static string Num(float value) => value.ToString("0.####", CultureInfo.InvariantCulture);

    private static string Camel(string name) =>
        string.IsNullOrEmpty(name) ? name : char.ToLowerInvariant(name[0]) + name[1..];
}

File: src/eQuantic.UI.Runtime/src/shared/theme-bridge.ts
/**
 * The client half of the SSR→client THEME BRIDGE: rehydrate the JSON the server serialized from the
 * app-selected `IAppTheme` (see C# `ThemeBridge.SerializeJson`) back into an `AppTheme` the runtime
 * swaps in via `setPhotonTheme` at boot. This is what lets an app pick Material — or a seeded theme —
 * for hydration: client re-renders then resolve the SAME colors/shape the server baked into the SSR
 * markup. The wire shape mirrors the generated `photonTheme`; the round-trip is pinned in vitest.
 */

import { ColorToken, TypeStyle, VariantColors, codeTokenColor } from './value-types';
import type { AppTheme, ShadowSpec } from './value-types';
import type { ColorValue } from './nodes';
import { DataPalette, DivergingScale, StatusScale } from './data-palette';

type RgbaTuple = readonly [number, number, number, number];
type TokenTuple = readonly [RgbaTuple, RgbaTuple];

/** The exact JSON `ThemeBridge.SerializeJson` emits into `window.__EQ_THEME__`. */
export interface ThemeData {
  surfaces: Record<string, TokenTuple>;
  disabledOpacity: number;
  variants: Record<string, readonly TokenTuple[]>; // [base, onBase, pressed, subtle, onSubtle]
  type: Record<string, readonly [number, number, string, number, number]>;
  elevations: ReadonlyArray<readonly [number, number, number, TokenTuple]>;
  shape: Record<string, number>;
  /** The data palette: series and sequential as token lists, diverging as [negative, midpoint,
   * positive], status as [good, warning, serious, critical]. */
  data: {
    series: readonly TokenTuple[];
    sequential: readonly TokenTuple[];
    diverging: readonly [TokenTuple, TokenTuple, TokenTuple];
    other: TokenTuple;
    status: readonly [TokenTuple, TokenTuple, TokenTuple, TokenTuple];
  };
}

const color = (t: RgbaTuple): ColorValue => ({ r: t[0], g: t[1], b: t[2], a: t[3] });
const token = (t: TokenTuple): ColorToken => new ColorToken(color(t[0]), color(t[1]));

/**
 * Rebuild an `AppTheme` from the serialized blob — the mirror of the generated `photonTheme` object:
 * surface tokens spread as top-level props, plus the four map-backed lookups (colors/type/elevation/
 * shape) with the same fallbacks the generator emits.
 */
export function materializeTheme(data: ThemeData): AppTheme {
  const surfaces: Record<string, ColorToken> = {};
  for (const key of Object.keys(data.surfaces)) surfaces[key] = token(data.surfaces[key]);

  const variants: Record<string, VariantColors> = {};
  for (const key of Object.keys(data.variants)) {
    const [base, onBase, pressed, subtle, onSubtle] = data.variants[key];
    variants[key] = new VariantColors(
      token(base),
      token(onBase),
      token(pressed),
      token(subtle),
      token(onSubtle),
    );
  }

  const typeScale: Record<string, TypeStyle> = {};
  for (const key of Object.keys(data.type)) {
    const [size, lineHeight, weight, tracking, maxScale] = data.type[key];
    typeScale[key] = new TypeStyle(size, lineHeight, weight, tracking, maxScale);
  }

  const elevations: ShadowSpec[] = data.elevations.map(([offsetY, blur, spread, c]) => ({
    offsetY,
    blur,
    spread,
    color: token(c),
  }));

  const palette = new DataPalette(
    data.data.series.map(token),
    data.data.sequential.map(token),
    new DivergingScale(
      token(data.data.diverging[0]),
      token(data.data.diverging[1]),
      token(data.data.diverging[2]),
    ),
    token(data.data.other),
    new StatusScale(
      token(data.data.status[0]),
      token(data.data.status[1]),
      token(data.data.status[2]),
      token(data.data.status[3]),
    ),
  );
  return {
    ...(surfaces as unknown as Pick<AppTheme, never>),
    data: palette,
    disabledOpacity: data.disabledOpacity,
    colors(variant: string): VariantColors {
      return variants[variant] ?? variants.primary;
    },
    type(role: string): TypeStyle {
      return typeScale[role] ?? typeScale.bodyL;
    },
    elevation(level: number): ShadowSpec {
      return elevations[Math.max(0, Math.min(5, Math.trunc(level)))];
    },
    code(kind: string): ColorToken {
      return codeTokenColor(this as unknown as AppTheme, kind);
    },
    shape(scale: string): number {
      return data.shape[scale] ?? data.shape.medium;
    },
  } as AppTheme;
}
